### PR TITLE
feat(settings): add reviewed bulk cleanup flows

### PR DESCRIPTION
**Pull request by JSONbored**
## Summary
- Adds reviewed bulk cleanup flows for merchants, categories, and account metadata.
- Keeps suggestions human-reviewed before applying changes.
- Adds safer validation around merge targets and reviewed website/domain input.

## Feature objective
Large imported or synced datasets can create duplicate merchants/categories and incomplete account/provider metadata. This PR adds reviewed, family-scoped bulk cleanup flows so users can clean up common data issues faster without auto-applying destructive changes.

## What changed
- Adds family-scoped bulk merge/update flows.
- Preserves provider merchant records while allowing local cleanup.
- Supports bulk website/domain updates for better logo and institution metadata.
- Adds merchant merge UI behavior so users choose either an existing target or create a new target, not both.
- Rejects invalid reviewed website input instead of silently dropping it.

## Screenshots

<a href="https://github.com/user-attachments/assets/1eb49b90-3e08-46df-90c2-f8a1a89c2ca4">
  Bulk update institution domains
</a>
<a href="https://github.com/user-attachments/assets/9fe25323-9cb7-4cf6-b3e5-5c47ff85f9f4">
  Bulk update merchant websites
</a>
<a href="https://github.com/user-attachments/assets/876c4c2d-808d-4998-a824-5f2e07db53f5">
  Merchant merge mutual exclusion
</a>
<a href="https://github.com/user-attachments/assets/b06a3038-8157-4bf3-b5e2-73784f98e68c">
  Category merge
</a>

## Why
Large imported or synced datasets often create duplicate merchants, categories, and incomplete account/provider metadata. Reviewed bulk cleanup makes Sure easier to maintain while keeping changes scoped, transactional, and reviewable before applying.

## Validation
- `bin/rails test test/controllers/accounts_controller_test.rb test/controllers/categories_controller_test.rb test/controllers/family_merchants_controller_test.rb test/models/account_test.rb test/models/merchant_test.rb`
- `bin/rails test test/controllers/family_merchants_controller_test.rb`
- `npm run lint`
- scoped `bin/rubocop`
- browser verification for the new cleanup dialogs/merchant merge behavior
- `git diff --check`

## Notes
- Screenshots added for the new user-facing flows.

## Summary by CodeRabbit

* **New Features**
  * Bulk edit institution domains (modal) with domain normalization, selection validation, and success/failure messaging
  * Bulk update merchant websites (modal) with domain extraction, logo refresh, and failure handling
  * Merge categories (modal) supporting create-or-select target, validations, and merge reporting
  * Client-side merge UI syncing for existing vs new targets

* **Other**
  * Deterministic default merchant color and centralized Brandfetch-based logo URLs
  * Accounts persist normalized institution domains

* **Tests**
  * Added integration and model tests for bulk flows, merges, and domain/logo normalization

### CODE REVIEW

File: app/controllers/family_merchants_controller.rb [hunk 1]
```diff
@@ -102,20 +102,49 @@ def merge
...
-    # Scope lookups to merchants valid for this family (FamilyMerchants + assigned ProviderMerchants)
-    valid_merchants = all_family_merchants
+    permitted_params = merchant_merge_params
 
-    target = valid_merchants.find_by(id: params[:target_id])
-    unless target
-      return redirect_to merge_family_merchants_path, alert: t(".target_not_found")
+    if conflicting_merge_target?(permitted_params)
+      return redirect_to merge_family_merchants_path, alert: t(".conflicting_target")
     end
```
Comment (jjmata) [inline app/controllers/family_merchants_controller.rb:143]: **Missing error handling — potential 500.**

`bulk_update_websites` has no `rescue ActiveRecord::RecordInvalid` block. If a `merchant.update!` call fails validation, the exception will bubble up and produce a 500. The two other bulk operations (`bulk_update_domains`, `perform_merge`) both rescue this; this one should too:

```ruby
redirect_to family_merchants_path, notice: t(".success", count: merchants.count)
rescue ActiveRecord::RecordInvalid => e
  redirect_to bulk_websites_family_merchants_path, alert: e.record.errors.full_messages.to_sentence
end
```

---
_Generated by [Claude Code](https://claude.ai/code)_

File: app/models/category/merger.rb [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        if target_budget_category
+          target_budget_category.update!(
+            budgeted_spending: target_budget_category.budgeted_spending + source_budget_category.budgeted_spending
+          )
+          source_budget_category.destroy!
+        else
+          source_budget_category.update!(category_id: target_category.id)
+        end
+      end
+    end
```
Comment (jjmata) [inline app/models/category/merger.rb:104]: **`reparent_subcategories` moves children to the target's *parent*, not to the target itself.**

When the target is a subcategory (has a `parent_id`), the source's children get reassigned to the target's *parent*, making them siblings of the target rather than children of it. For example: merging `Food > Coffee` (source) into `Food > Restaurants` (target) would move `Coffee`'s subcategories up to `Food`, not down into `Restaurants`.

Is this intentional? If so, it's worth documenting. If not, the condition should likely be:

```ruby
def reparent_subcategories(source)
  family.categories.where(parent_id: source.id).where.not(id: target_category.id).update_all(parent_id: target_category.id)
end
```

Note also that `dependent: :nullify` is set on `Category#subcategories`, so `source.destroy!` will nullify `parent_id` on any subcategories not explicitly reparented here — you'd want to ensure the `update_all` runs before `destroy!`, which it does in the current code.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: app/views/accounts/bulk_domains.html.erb [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+            name: :institution_domain,
+            required: true,
+            label: t(".domain_label"),
+            placeholder: t(".domain_placeholder") %>
+
+      <div class="space-y-2">
+        <p class="text-sm font-medium text-primary"><%= t(".accounts_label") %></p>
+        <div class="max-h-64 overflow-y-auto space-y-1 border border-secondary rounded-lg p-2">
+          <% @accounts.each do |account| %>
+            <label class="flex items-center gap-2 p-2 hover:bg-surface-hover rounded cursor-pointer">
```
Comment (jjmata) [inline app/views/accounts/bulk_domains.html.erb:15]: **Design system violation: `border-gray-300`.**

Per the project conventions, raw Tailwind primitives should not be used directly — functional tokens from the design system are required. The same raw class appears in `categories/merge.html.erb` and `family_merchants/bulk_websites.html.erb` as well.

Existing usage elsewhere in the app (e.g. `family_merchants/merge.html.erb`) doesn't use `border-gray-300` on checkboxes. The simplest fix is to drop the border class entirely and rely on the browser/OS default checkbox styling, or use whichever design-system utility is appropriate (e.g. `border-secondary`).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: app/controllers/categories_controller.rb [hunk 4]
```diff
@@ -89,4 +138,24 @@ def set_transaction
...
+    end
+
+    def conflicting_merge_target?(permitted_params)
+      permitted_params[:target_id].present? && permitted_params[:new_target_name].present?
+    end
+
+    def merge_target_category(permitted_params)
+      if permitted_params[:new_target_name].present?
+        Current.family.categories.create!(
+          name: permitted_params[:new_target_name],
```
Comment (jjmata) [inline app/controllers/categories_controller.rb:146]: **`COLORS.sample` fallback picks a random color non-deterministically.**

The form pre-populates the color picker with `Category::COLORS.first`, so the user always sees a chosen color. But if the field arrives empty server-side (e.g. a crafted request), the saved color would be random, producing a mismatch from what was shown.

A safer fallback is the same default the form uses:

```ruby
color: permitted_params[:new_target_color].presence || Category::COLORS.first,
```

The same issue exists in `merge_target_merchant` for `FamilyMerchant::COLORS.sample`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: app/controllers/categories_controller.rb [hunk 5]
```diff
@@ -67,6 +71,51 @@ def bootstrap
...
+    unless target
+      return redirect_to merge_categories_path, alert: t(".target_not_found")
+    end
+
+    if merge_succeeded
+      redirect_to categories_path, notice: t(".success", count: merger.merged_count)
+    else
+      redirect_to merge_categories_path, alert: t(".no_categories_selected")
+    end
+  rescue Category::Merger::UnauthorizedCategoryError => e
```
Comment (jjmata) [inline app/controllers/categories_controller.rb:102]: **Controller `perform_merge` is harder to follow than necessary.**

Initializing `target`, `merger`, and `merge_succeeded` as `nil`/`false` outside the transaction block, then checking them again after the block, is a pattern that relies on Ruby variable mutation across a `raise ActiveRecord::Rollback` — which is non-obvious and could surprise maintainers who don't know that `Rollback` silently absorbs the exception rather than propagating it.

A cleaner approach: move the redirect logic so that all branching happens *inside* the transaction for the happy path, and only rescue/redirect outside. For example:

```ruby
def perform_merge
  permitted_params = category_merge_params
  return redirect_to merge_categories_path, alert: t(".conflicting_target") if conflicting_merge_target?(permitted_params)

  sources = Current.family.categories.where(id: permitted_params[:source_ids])
  return redirect_to merge_categories_path, alert: t(".invalid_categories") unless sources.any?

  ActiveRecord::Base.transaction do
    target = merge_target_category(permitted_params) or raise ActiveRecord::Rollback
    merger = Category::Merger.new(family: Current.family, target_category: target, source_categories: sources)
    merger.merge! or raise ActiveRecord::Rollback
    # if we reach here, commit and redirect
    return redirect_to categories_path, notice: t(".success", count: merger.merged_count)
  end

  redirect_to merge_categories_path, alert: t(".no_categories_selected")
rescue Category::Merger::UnauthorizedCategoryError => e
  redirect_to merge_categories_path, alert: e.message
rescue ActiveRecord::RecordInvalid => e
  redirect_to merge_categories_path, alert: e.record.errors.full_messages.to_sentence
end
```

This makes the control flow explicit: return from inside the transaction on success, fall through to an alert redirect if the transaction was rolled back.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: app/models/category/merger.rb [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    sources = Array(source_categories)
+    sources.each { |category| validate_category_belongs_to_family!(category, "Source category '#{category.name}'") }
+
+    @source_categories = sources.reject { |category| category.id == target_category.id }
+    validate_hierarchy!
+  end
+
+  def merge!
+    return false if source_categories.empty?
+
```
Comment (jjmata) [inline app/models/category/merger.rb:23]: **`Category::Merger#merge!` opens its own transaction, but the controller already wraps the call in one.**

Nested `ActiveRecord::Base.transaction` blocks use savepoints in PostgreSQL (and are no-ops in SQLite unless `requires_new: true` is set). The inner transaction here doesn't add real isolation — the outer transaction in the controller is what ultimately governs rollback. This double-wrapping is a code smell and could cause confusion if the behaviour ever needs to differ between the two.

Consider either:
- having the controller manage the transaction and `merge!` operate without one (caller's responsibility), or
- having `merge!` own the transaction and the controller not wrap it.

Since `Category::Merger` is a domain object that may be called from contexts other than this controller, having it own the transaction (option 2) is probably the cleaner long-term design — but then the controller shouldn't open its own outer transaction.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: app/views/categories/merge.html.erb [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+  <% dialog.with_body do %>
+    <%= styled_form_with url: perform_merge_categories_path, method: :post, class: "space-y-4 max-h-[70vh] overflow-y-auto pr-1" do |f| %>
+      <%= f.collection_select :target_id,
+            @categories,
+            :id, :name_with_parent,
+            { prompt: t(".select_target"), label: t(".target_label") },
+            {} %>
+
+      <div class="space-y-3 rounded-lg border border-secondary p-3">
+        <p class="text-sm font-medium text-primary"><%= t(".new_target_label") %></p>
```
Comment (jjmata) [inline app/views/categories/merge.html.erb:12]: **Same `@categories` list is used for both the target dropdown and the sources checklist — a user can select the same category as both.**

`Category::Merger` silently filters out the target from the sources (`sources.reject { |c| c.id == target_category.id }`), so there's no data hazard. But from a UX perspective, a user who selects "Dining" as target and also checks "Dining" as a source will see it silently ignored with no feedback — the success message will report one fewer merge than they expected.

A small JS enhancement (or even just a note in the hint text) to clarify that selecting a category as target automatically excludes it from the source list would prevent confusion.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by jjmata — COMMENTED** [review]
## Review: feat(settings) — reviewed bulk cleanup flows

Good overall direction. The domain-extraction consolidation (`Merchant.extract_domain` / `brandfetch_logo_url_for`) cleans up a real duplication, the scoping is correct throughout, and the test coverage for the happy/sad paths is solid.

A few things need attention before merge:

**Must fix**
- `bulk_update_websites` is missing a `rescue ActiveRecord::RecordInvalid` block — a validation failure will 500 (see inline comment).
- `reparent_subcategories` moves orphaned source children to the *target's parent*, not to the target itself — needs a decision on intended behaviour and likely a fix (see inline comment).

**Should fix**
- Three templates use `border-gray-300` directly, violating the project's design-system token rule.
- `COLORS.sample` fallback is non-deterministic; use `COLORS.first` to match the form default.

**Nits / suggestions**
- `perform_merge` in `CategoriesController` uses a confusing variable-mutation-across-rollback pattern; a simpler control flow is suggested inline.
- `Category::Merger#merge!` wraps itself in a transaction that's already opened by the controller — pick one owner.
- The target dropdown and sources checklist share the same `@categories` list with no UX indication that selecting the same item in both is a no-op.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by jjmata — CHANGES_REQUESTED** [review]

**Review by jjmata — COMMENTED** [review]
## Re-review — all previous issues addressed, no new concerns

All 7 issues from the prior review have been fixed, and the additional fixes prompted by CodeRabbit are solid. Summary:

**Previously flagged — confirmed resolved:**

1. **Missing `rescue ActiveRecord::RecordInvalid` in `bulk_update_websites`** ✅ — rescue added with a `t(".failure")` redirect; locale key included.

2. **`reparent_subcategories` moved children to the target's parent instead of the target** ✅ — now uses `parent_id: target_category.id`. A new `validate_reparenting!` guard also prevents merging a source-with-subcategories into a subcategory target (which would have created an illegal 3-level hierarchy via `update_all`). Two new controller tests confirm the corrected direction and the rejection case.

3. **`border-gray-300` design system violation in three templates** ✅ — replaced with `border-secondary` throughout.

4. **`COLORS.sample` non-deterministic fallback** ✅ — `Category::COLORS.first` and `FamilyMerchant::COLORS.first` used consistently in both controllers and in `FamilyMerchant#set_default_color`.

5. **`perform_merge` controller complexity** ✅ — refactored into a clean `merge_categories!` helper that raises typed sentinel exceptions (`MergeTargetNotFound`, `EmptyCategoryMerge`), giving explicit control flow without the variable-mutation-across-rollback pattern.

6. **Nested transaction in `Category::Merger#merge!`** ✅ — the inner `Category.transaction` was removed from `merge!`; the outer transaction in `merge_categories!` owns it. The transaction boundary is now unambiguous.

7. **Target/source same-list UX** ✅ — `target_selected_as_source?` guard added server-side with a redirect and alert; hint text updated to "Do not select the target as a source." Pragmatic fix without requiring JS.

**Notable improvements in subsequent commits (CodeRabbit-prompted):**
- `FamilyMerchant#set_default_color` now only sets a default when `color.blank?` — user-specified colors are preserved on the new-target merge form.
- `ProviderMerchant#generate_logo_url_from_website!` and `FamilyMerchant#generate_logo_url_from_website` now clear stale logos when Brandfetch is unavailable (removed the `brand_fetch_client_id` guard; `brandfetch_logo_url_for` handles the nil return internally).
- `Account.normalize_institution_domain` now delegates to `Merchant.extract_domain`, eliminating the separate implementation and the raw-URL fallback. Extra test coverage for the nil case (`"https://bad host"`) added.
- `ActiveRecord::RecordNotDestroyed` is now caught alongside `RecordInvalid` in `categories#perform_merge`.

**No new issues found.** The code is well-structured. The one open thread from `chatgpt-codex-connector` (P1 — settings layout on `bulk_domains`) is outdated; `bulk_domains` correctly uses `render layout: false` in the current code.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by jjmata** [comment]
Please include screenshots of new functionality and high-level description of feature objective, @JSONbored. 🙏

**Comment by JSONbored** [comment]
> Please include screenshots of new functionality and high-level description of feature objective, @JSONbored. 🙏

Thanks, added the high-level feature objective and screenshots to the PR description.

I also attached screenshots for the new visible flows: bulk institution-domain cleanup, bulk merchant website cleanup, merchant merge target selection/new-target creation, and category merge.

Latest validation is listed in the PR description as well.

**Review by sure-design — CHANGES_REQUESTED** [review]
## Re-review — independent read of current code

I read the code directly (not relying on prior review comments) and confirmed what was fixed vs. what remains.

---

### Previously flagged open threads that are now addressed in the current code

Several `jjmata` threads are still marked unresolved but the code already has the fix:

- **Missing rescue in `bulk_update_websites`** — The `rescue ActiveRecord::RecordInvalid` block is present in the current file. Thread can be closed.
- **`border-gray-300` design system violation** — All three views (`bulk_domains`, `categories/merge`, `bulk_websites`) now use `border-secondary`. Thread can be closed.
- **`COLORS.sample` non-deterministic fallback** — Categories use `Category::COLORS.first`; merchants use `FamilyMerchant.default_color` (= `COLORS.first`). Thread can be closed.
- **Double-wrapping transactions** — `Category::Merger#merge!` has no transaction of its own. Only the controller's `Category.transaction` in `merge_categories!` exists. The concern is moot. Thread can be closed.
- **`reparent_subcategories` moves to target's parent** — The code correctly uses `update_all(parent_id: target_category.id)` (children go under the target, not under the target's parent). jjmata's suggested fix is identical to what's already in the code. Thread can be closed.

---

### Real bug — one new issue

**`FamilyMerchantsController#perform_merge` can create an orphaned target merchant**

This is the same atomicity issue that CodeRabbit flagged for `categories_controller` (commit `0a94adb`) — and that was fixed by wrapping everything in `Category.transaction` via `merge_categories!`. That fix was never applied to the merchant equivalent.

Current flow in `perform_merge`:

```ruby
target = merge_target_merchant(valid_merchants, permitted_params)  # calls create! — committed immediately
# ...
merger = Merchant::Merger.new(family: Current.family, target_merchant: target, source_merchants: sources)

if merger.merge!
  redirect_to family_merchants_path, ...
else
  redirect_to merge_family_merchants_path, ...
end
rescue Merchant::Merger::UnauthorizedMerchantError => e
  redirect_to merge_family_merchants_path, alert: e.message   # target merchant already in DB, orphaned
rescue ActiveRecord::RecordInvalid => e
  redirect_to merge_family_merchants_path, alert: ...          # same problem
```

If `merger.merge!` raises `ActiveRecord::RecordInvalid` (e.g., a validation fires while reassigning transactions) or `UnauthorizedMerchantError`, the rescue redirects back with an alert — but the newly created target `FamilyMerchant` is already committed and persists as an orphan with no transactions assigned to it.

**Fix** — same pattern as categories: wrap target creation + merge in a transaction:

```ruby
def perform_merge
  permitted_params = merchant_merge_params
  return redirect_to merge_family_merchants_path, alert: t(".conflicting_target") if conflicting_merge_target?(permitted_params)

  valid_merchants = all_family_merchants
  sources = valid_merchants.where(id: permitted_params[:source_ids])
  return redirect_to merge_family_merchants_path, alert: t(".invalid_merchants") unless sources.any?

  merger = nil
  merge_succeeded = false

  Merchant.transaction do
    target = merge_target_merchant(valid_merchants, permitted_params) ||
             (return redirect_to merge_family_merchants_path, alert: t(".target_not_found"))
    merger = Merchant::Merger.new(family: Current.family, target_merchant: target, source_merchants: sources)
    merge_succeeded = merger.merge!
  end

  if merge_succeeded
    redirect_to family_merchants_path, notice: t(".success", count: merger.merged_count)
  else
    redirect_to merge_family_merchants_path, alert: t(".no_merchants_selected")
  end
rescue Merchant::Merger::UnauthorizedMerchantError => e
  redirect_to merge_family_merchants_path, alert: e.message
rescue InvalidMerchantWebsite
  redirect_to merge_family_merchants_path, alert: t(".invalid_website")
rescue ActiveRecord::RecordInvalid => e
  redirect_to merge_family_merchants_path, alert: e.record.errors.full_messages.to_sentence
end
```

---

### Minor issues

**`Account#logo_url` redundantly normalizes on every call**

`institution_domain` is already normalized on save by `before_validation :normalize_institution_domain_value`. The extra `normalize_institution_domain` call in `logo_url` is a no-op (the value is already canonical or `nil`), but it runs `URI.parse` + a regex on every `logo_url` invocation. Replace with the raw attribute:

```ruby
def logo_url
  domain = institution_domain  # already normalized by before_validation
  if domain.present? && Setting.brand_fetch_client_id.present?
    # ...
  end
end
```

**`brandfetch_logo_url_for` test validates an impossible input path**

```ruby
Merchant.stubs(:extract_domain).returns("example.com/path")
```

The real `extract_domain` validates with `/\A[a-z0-9.-]+\.[a-z0-9-]+\z/i`, which rejects forward slashes — so `"example.com/path"` can never be returned in production. The test confirms the encoding helper works, but it only exercises a code path that doesn't exist without the stub. Consider either testing with a value that `extract_domain` can actually return, or noting that encoding is a purely defensive layer.

**`FamilyMerchantsController#merge` renders in the default layout, not the settings layout**

`CategoriesController#merge` explicitly does `render layout: "settings"`. `FamilyMerchantsController#merge` has no `render` call, so it falls back to the application layout on direct navigation. Accessed via `frame: :modal` it works fine (Turbo extracts the frame), but the experience is inconsistent if a user lands directly on that URL.

---

_Generated by [Claude Code](https://claude.ai/code)_

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by jjmata — CHANGES_REQUESTED** [review]
One final pass?  See last Claude Code comment ...

**Comment by sure-design** [comment]
The `fix(settings): reuse merchant logo helpers` commit (May 6) looks correct — consolidating the Brandfetch URL construction into the model is the right call.

A couple of things to verify before jjmata's code review:

- **Cross-family data leak in `Merchant::Merger`**: the bulk merge flow touches provider merchant records. Confirm the family-scoping guard is tight enough that a provider merchant shared across families can't be merged/redirected by one family's user.
- **`valid_domain?` rejection path**: the PR rejects invalid website input rather than silently dropping it. Make sure the error message is surfaced back to the user in the form response, not just dropped silently server-side.

Otherwise the description and screenshots are now in good shape — ready for a substantive code review from @jjmata.

---
_Code review note from [Claude Code](https://claude.ai/code)_

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by sure-design** [comment]
New commits pushed 2026-05-07. The original 7-issue review round was fully resolved and confirmed by @jjmata on 2026-05-03. The remaining open item from the 2026-05-06 re-review is the atomicity bug in `FamilyMerchantsController#perform_merge`: `merge_target_merchant` calls `create!` (committed immediately) before `Merchant::Merger#merge!` runs, so a merge failure leaves an orphaned target merchant. The parallel fix for `CategoriesController` (wrapping in `Category.transaction`) was not applied to the merchant flow. Please confirm the new commits address this.

_Generated by [Claude Code](https://claude.ai/code)_

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by JSONbored** [comment]
> New commits pushed 2026-05-07. The original 7-issue review round was fully resolved and confirmed by @jjmata on 2026-05-03. The remaining open item from the 2026-05-06 re-review is the atomicity bug in `FamilyMerchantsController#perform_merge`: `merge_target_merchant` calls `create!` (committed immediately) before `Merchant::Merger#merge!` runs, so a merge failure leaves an orphaned target merchant. The parallel fix for `CategoriesController` (wrapping in `Category.transaction`) was not applied to the merchant flow. Please confirm the new commits address this.
> 
> _Generated by [Claude Code](https://claude.ai/code)_
> 
> _Generated by [Claude Code](https://claude.ai/code)_

Thanks, I verified the current branch against those points.

The cross-family provider merchant path is covered now: `Merchant::Merger` validates target/source merchants against the current family’s family-owned or assigned merchant IDs, only reassigns `family.transactions.where(merchant_id: source.id)`, and only destroys `FamilyMerchant` sources. There is also a regression test proving a shared `ProviderMerchant` merge only reassigns the current family’s transaction while the other family’s transaction still points to the shared provider merchant.

The invalid website path is also visible to the user: invalid reviewed `new_target_website_url` raises `InvalidMerchantWebsite`, redirects back to the merge form, and sets the localized `invalid_website` flash alert. The controller test asserts that flash path and that no target merchant is created.

I also confirmed the merchant merge atomicity issue is addressed: target creation and merge execution now run inside `Merchant.transaction`, `Merchant::Merger#merge!` is internally transaction-wrapped, and there is rollback coverage proving a failed merge does not leave an orphaned new target merchant.

Local targeted validation passed:
- `bin/rails test test/controllers/family_merchants_controller_test.rb test/models/merchant_test.rb test/controllers/accounts_controller_test.rb`
- `npm run lint`
- scoped `bin/rubocop`
- `git diff --check`
